src/main/webapp/app/exam/participate/exam-participation.service.ts

```diff
@@ -1,9 +1,145 @@
+                    this.studentExam = studentExam;
+                }),
+            );
+        }
+    }
+
+    private getLocalStorageKeyForStudentExam(courseId: number, examId: number) {
+        const prefix = 'artemis_student_exam';
+        return `${prefix}_${courseId}_${examId}`;
+    }
+
+    /**
+     * Retrieves a {@link StudentExam} from server
+     */
+    private getStudentExamFromServer(courseId: number, examId: number): Observable<StudentExam> {
+        return this.httpClient.get<StudentExam>(this.resourceUrl);
+    }
+
+    /**
+     * start AutoSaveTimer
+     */
+    public startAutoSaveTimer(): void {
+        // auto save of submission if there are changes
+        this.autoSaveInterval = window.setInterval(() => {
+            this.synchronizeSubmissionsWithServer();
+        }, 1000 * this.autoSaveTime);
+    }
+
+    /**
+     * creates submissions for all submissions in SubmissionSaveList
+     */
+    private synchronizeSubmissionsWithServer() {
+        // TODO: handle synchronization properly
+        this.submissionSaveList.forEach((submission) => {
+            switch (submission.participation.exercise?.type) {
+                case ExerciseType.TEXT:
+                    this.textSubmissionService.create(submission as TextSubmission, submission.participation.exercise?.id);
+                    break;
+                case ExerciseType.FILE_UPLOAD:
+                    return this.fileUploadSubmissionService;
+                case ExerciseType.MODELING:
+                    this.modelingSubmissionService.create(submission as ModelingSubmission, submission.participation.exercise.id).subscribe(
+                        (response) => {
+                            submission = response.body!;
+                            submission.participation.submissions = [submission];
+                            this.onSaveSuccess();
+                        },
+                        () => this.onSaveError(),
+                    );
+                    break;
+                case ExerciseType.PROGRAMMING:
+                    return this.programmingSubmissionService;
+                case ExerciseType.QUIZ:
+                    // TODO find submissionService
+                    return null;
+            }
+        });
+    }
+
+    private onSaveSuccess() {
+        console.log('saved');
+    }
+
+    private onSaveError() {
+        console.log('error while saving');
+    }
+
+    /**
+     * Updates StudentExam on server
+     * @param studentExam
+     */
+    createSubmission(submission: Submission, courseId: number, examId: number, exerciseId: number) {
+        // TODO: Add to updates
+
+        // update immediately in localStorage and online every 60seconds
+        this.localStorageService.store(this.getLocalStorageKeyForStudentExam(courseId, examId), JSON.stringify(this.studentExam));
+    }
+
+    getLatestSubmissionForParticipation(participationId: number): Observable<Submission | undefined> {
```

Comment:
I don't think this method should be necessary.
When the user clicks on start in the exam, the studentExam for the user should be downloaded with all exercises, participations and (for text, modeling, quiz) an initial empty submission.
(In the 2nd iteration, we can develop a fallback solution, in case one participation does not exist, then we would invoke startExercise on the server, when the student navigates into the exercise for the first time)

During the exam, there should be no need to load these objects again from the server

Reply:
You are right. One question though: What happens when the student switches the browser for whatever reason? Will the most current submissions be downloaded with the studentExam?

Reply:
Yes, then the student needs to "start" the exam again and the latest (synced) submission will be downloaded in StudentExam -> Exercises -> Participations -> Submissions